File: iOS7bookExamples/bk2ch24p842downloader/ch37p1099downloader/MyDownloader.m


#import "MyDownloader.h"

@interface MyDownloader() <NSURLSessionDownloadDelegate>
@property (nonatomic, strong) NSURLSession* session;
@property (nonatomic, strong) NSOperationQueue* q;
@end

@implementation MyDownloader

#define which 2 // use 2 to move delegate methods onto a background thread

- (id) initWithConfiguration: (NSURLSessionConfiguration*) config {
    self = [super init];
    if (self) {
        self->_q = [NSOperationQueue new];
#if (which == 2)
        NSURLSession* session = [NSURLSession sessionWithConfiguration:config delegate:self delegateQueue:self->_q];
#else
        NSURLSession* session = [NSURLSession sessionWithConfiguration:config delegate:self delegateQueue:[NSOperationQueue mainQueue]];
#endif
        self->_session = session;
    }
    return self;
}

- (NSURLSessionTask*) download:(NSString*)s completionHandler:(void(^)(NSURL* url))ch {
    NSURL* url = [NSURL URLWithString:s];
    NSMutableURLRequest* req = [NSMutableURLRequest requestWithURL:url];
    [NSURLProtocol setProperty:ch forKey:@"ch" inRequest:req];
    NSURLSessionDownloadTask* task = [self.session downloadTaskWithRequest:req];
    [task resume];
    return task;
}

-(void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didWriteData:(int64_t)bytesWritten totalBytesWritten:(int64_t)totalBytesWritten totalBytesExpectedToWrite:(int64_t)totalBytesExpectedToWrite {
    NSLog(@"downloaded %d%%", (int)(100.0*totalBytesWritten/totalBytesExpectedToWrite));
}

-(void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didResumeAtOffset:(int64_t)fileOffset expectedTotalBytes:(int64_t)expectedTotalBytes {
    // unused in this example
    NSLog(@"%@", @"did resume");
}

-(void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didFinishDownloadingToURL:(NSURL *)location {
    NSURLRequest* req = downloadTask.originalRequest;
    void(^ch)(NSURL* url) = [NSURLProtocol propertyForKey:@"ch" inRequest:req];
    NSHTTPURLResponse* response = (NSHTTPURLResponse*)downloadTask.response;
    NSInteger stat = response.statusCode;
    NSLog(@"status %ld", (long)stat);
    NSURL* url = nil;
    if (stat == 200) {
        url = location;
        NSLog(@"downloaded %@", req.URL.lastPathComponent);
    }
#if (which == 2)
    dispatch_sync(dispatch_get_main_queue(), ^{
        ch(url);
    });
#else
    ch(url);
#endif
}

- (void) cancelAllTasks {
    [self.session invalidateAndCancel];
}

- (void) dealloc {
    NSLog(@"%@", @"MyDownloader dealloc");
}



@end

File: bk2ch24p842downloaderOld/ch37p1099downloader/MyDownloader.swift


import UIKit

let isMain = false // try false to move delegate methods onto a background thread

typealias MyDownloaderCompletion = (URL?) -> ()

// decided to give up the URLProtocol approach
// ... and just keep an array of completion handlers in the delegate

class MyDownloader: NSObject {
    let config : URLSessionConfiguration
    let q = OperationQueue()
    lazy var session : URLSession = {
        let queue = (isMain ? .main : self.q)
        return URLSession(configuration:self.config, delegate:MyDownloaderDelegate(), delegateQueue:queue)
    }()
    init(configuration config:URLSessionConfiguration) {
        self.config = config
        super.init()
    }
    
    @discardableResult
    func download(url:URL, completionHandler ch : @escaping MyDownloaderCompletion) -> URLSessionTask {
        let req = NSMutableURLRequest(url:url)
        URLProtocol.setProperty(ch, forKey:"ch", in:req)
        let task = self.session.downloadTask(with:req as URLRequest)
        task.resume()
        return task
    }
    
    private class MyDownloaderDelegate : NSObject, URLSessionDownloadDelegate {
        func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didFinishDownloadingTo url: URL) {
            let req = downloadTask.originalRequest!
            let ch = URLProtocol.property(forKey:"ch", in:req) as! MyDownloaderCompletion
            if isMain {
                ch(url)
            } else {
                DispatchQueue.main.sync {
                    ch(url)
                }
            }
        }
        deinit {
            print("farewell from Delegate")
        }
    }

    
//    func cancelAllTasks() {
//        self.session.invalidateAndCancel()
//    }
    
    deinit {
        print("farewell from MyDownloader")
        self.session.invalidateAndCancel()
    }
    
}

